File: backend/src/auth/auth.controller.ts
import {
  Controller,
  Get,
  Post,
  Body,
  Req,
  Res,
  HttpCode,
  HttpStatus,
  UseGuards,
} from '@nestjs/common';
import type { Request, Response } from 'express';
import { AuthService } from './auth.service';
import { JwtAuthGuard } from './jwt-auth.guard';
import { loginSchema } from './auth.validation';

const COOKIE_NAME = 'admin_token';
const EIGHT_HOURS_MS = 8 * 60 * 60 * 1000;

@Controller('auth')
export class AuthController {
  constructor(private readonly authService: AuthService) {}

  @Post('login')
  @HttpCode(HttpStatus.OK)
  async login(
    @Body() body: unknown,
    @Res({ passthrough: true }) res: Response,
  ) {
    const result = loginSchema.safeParse(body);
    if (!result.success) {
      const errors = result.error.issues.map((issue) => ({
        field: issue.path.join('.'),
        message: issue.message,
      }));
      return { statusCode: 400, message: 'Validation failed', errors };
    }

    const { access_token } = await this.authService.login(
      result.data.email,
      result.data.password,
    );

    res.cookie(COOKIE_NAME, access_token, {
      httpOnly: true,
      secure: process.env.NODE_ENV === 'production',
      sameSite: 'strict',
      maxAge: EIGHT_HOURS_MS,
      path: '/',
    });

    return { ok: true };
  }

  @Post('logout')
  @HttpCode(HttpStatus.OK)
  logout(@Res({ passthrough: true }) res: Response) {
    res.clearCookie(COOKIE_NAME, { path: '/' });
    return { ok: true };
  }

  @UseGuards(JwtAuthGuard)
  @Get('me')
  me(@Req() req: Request & { user: { sub: string; email: string } }) {
    return { id: req.user.sub, email: req.user.email };
  }
}

File: backend-dotnet/src/HealthTech.Api/Controllers/AuthController.cs
using System.Security.Claims;
using FluentValidation;
using HealthTech.Api.Services;
using HealthTech.Shared.Dtos;
using Microsoft.AspNetCore.Authorization;
using Microsoft.AspNetCore.Mvc;
using Microsoft.AspNetCore.RateLimiting;

namespace HealthTech.Api.Controllers;

[ApiController]
[Route("api/[controller]")]
public class AuthController : ControllerBase
{
    private const string CookieName = "admin_token";
    private static readonly TimeSpan CookieMaxAge = TimeSpan.FromHours(8);

    private readonly IAuthService _authService;
    private readonly IValidator<LoginRequest> _loginValidator;
    private readonly IConfiguration _configuration;

    public AuthController(IAuthService authService, IValidator<LoginRequest> loginValidator, IConfiguration configuration)
    {
        _authService = authService;
        _loginValidator = loginValidator;
        _configuration = configuration;
    }

    private bool IsCookieSecure() =>
        _configuration["COOKIE_SECURE"] != "false" &&
        Environment.GetEnvironmentVariable("COOKIE_SECURE") != "false";

    [HttpPost("login")]
    [EnableRateLimiting("login")]
    public async Task<IActionResult> Login([FromBody] LoginRequest request)
    {
        var validation = await _loginValidator.ValidateAsync(request);
        if (!validation.IsValid)
        {
            return BadRequest(new
            {
                message = "Validation failed",
                errors = validation.Errors.Select(e => new
                {
                    field = e.PropertyName.ToLowerInvariant(),
                    message = e.ErrorMessage,
                }),
            });
        }

        var result = await _authService.LoginAsync(request.Email, request.Password);
        if (result is null)
            return Unauthorized(new { message = "Invalid email or password" });

        var cookieSecure = IsCookieSecure();

        Response.Cookies.Append(CookieName, result.Value.token, new CookieOptions
        {
            HttpOnly = true,
            Secure = cookieSecure,
            SameSite = SameSiteMode.Strict,
            MaxAge = CookieMaxAge,
            Path = "/",
        });

        return Ok(new AuthResponse { Ok = true });
    }

    [HttpPost("logout")]
    [Authorize]
    public IActionResult Logout()
    {
        var cookieSecure = IsCookieSecure();

        Response.Cookies.Delete(CookieName, new CookieOptions
        {
            HttpOnly = true,
            Secure = cookieSecure,
            SameSite = SameSiteMode.Strict,
            Path = "/",
        });

        return Ok(new AuthResponse { Ok = true });
    }

    [HttpGet("me")]
    [Authorize]
    public IActionResult Me()
    {
        var sub = User.FindFirstValue(System.IdentityModel.Tokens.Jwt.JwtRegisteredClaimNames.Sub)
            ?? User.FindFirstValue(ClaimTypes.NameIdentifier);
        var email = User.FindFirstValue(System.IdentityModel.Tokens.Jwt.JwtRegisteredClaimNames.Email)
            ?? User.FindFirstValue(ClaimTypes.Email);

        if (sub is null || email is null)
            return Unauthorized(new { message = "Invalid token claims" });

        return Ok(new MeResponse { Id = Guid.Parse(sub), Email = email });
    }
}
